### ISSUE TITLE
Linter wants previously declared variable destructured during assignment

### ISSUE THREAD

**Issue by krishnaglick** [eslint/eslint#9661]
**Tell us about your environment**
Windows 10

* **ESLint Version:** v4.12.0
* **Node Version:** v8.9.1
* **npm Version:** 5.5.

**What parser (default, Babel-ESLint, etc.) are you using?**
eslint

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```
{
  "env": {
    "es6": true,
    "node": true,
    "browser": true
  },
  "parserOptions": {
    "sourceType": "module",
    "ecmaVersion": 8
  },
  "rules": {
    "prefer-destructuring": ["error", {
      "VariableDeclarator": {
        "array": true,
        "object": true
      },
      "AssignmentExpression": {
        "array": true,
        "object": true
      }
    }, {
      "enforceForRenamedProperties": false
    }]
  }
}

```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```
function potato(b) {
  let a;
  if(b) {
    a = b.a;
  }

  doThing(a); //Where a being undefined or not affects the downstream.
}

```

```
eslint .
```

**What did you expect to happen?**
No error or warning.

**What actually happened? Please include the actual, raw output from ESLint.**
```
..path...\test.js
  5:5  error  Use object destructuring  prefer-destructuring

✖ 1 problem (1 error, 0 warnings)
```

**Comment by platinumazure** [comment]
Maybe I'm missing something, but can't you just do a destructuring assignment outside of the declaration?

```js
function potato(b) {
  let a;
  if(b) {
    {a} = b;
  }

  doThing(a); //Where a being undefined or not affects the downstream.
}
```

**Comment by krishnaglick** [comment]
Altered js to:
test.js
```

const doThing = console.log;

function potato(b) {
  let a;
  if(b) {
    {a} = b;
  }

  doThing(a); //Where a being undefined or not affects the downstream.
}

potato();
```

And ran with `node test.js`

```
\test.js:7
    {a} = b;
        ^

SyntaxError: Unexpected token =
```

**Comment by platinumazure** [comment]
Sorry, I forget that destructuring object assignments need parentheses:

```js
({a} = b);
```

**Comment by krishnaglick** [comment]
Ah, that fixed it. Is there a linting rule to yell about that, or more a `don't be dumb` kinda thing? :)

**Comment by platinumazure** [comment]
Unfortunately, since that's a parsing error, the lint rules can't even run
at that point. I'm not aware of a good way to catch that and report a
user-friendly message explaining the need for parentheses... Sorry.
